ui/src/shared/utils/fromFlux.ts

```diff
@@ -0,0 +1,262 @@
+                 |                   |          |               1.0 |     true
+                 |                   |          |               2.0 |     true
+                 |                   |          |               3.0 |     true
+
+    The `#group`, `#datatype`, and `#default` [annotations][1] are expected to be
+    in the input Flux CSV.
+    Values are coerced into appropriate JavaScript types based on the Flux
+    `#datatype` annotation for the table
+    The `Table` stores a `key` for each column which is seperate from the column
+    `name`. If multiple Flux tables have the same column but with different
+    types, they will be distinguished by different keys in the resulting `Table`;
+    otherwise the `key` and `name` for each column in the result table will be
+    identical.
+
+    [0]: https://github.com/influxdata/flux/blob/master/docs/SPEC.md#csv
+    [1]: https://github.com/influxdata/flux/blob/master/docs/SPEC.md#annotations
+    */
+
+export default function fromFlux(csv) {
+  /*
+         A Flux CSV response can contain multiple CSV files each joined by a newline.
+         See https://github.com/influxdata/flux/blob/master/docs/SPEC.md#multiple-tables.
+
+         Split the response into separate chunks whenever we encounter:
+             1. A newline
+             2. Followed by any amount of whitespace
+             3. Followed by a newline
+             4. Followed by a `#` character
+         The last condition is [necessary][0] for handling CSV responses with
+         values containing newlines.
+
+         [0]: https://github.com/influxdata/influxdb/issues/15017
+     */
+  const output = {},
+    groupKey = {},
+    names = {}
+
+  const chunks = [],
+    regerz = /\n\s*\n#/g
+  let match,
+    lastRange = 0
+
+  while ((match = regerz.exec(csv)) !== null) {
+    chunks.push({
+      start: lastRange,
+      stop: match.index,
+    })
+    lastRange = match.index + match[0].length - 1
+  }
+  chunks.push({
+    start: lastRange,
+    stop: csv.length,
+  })
+
+  let runningTotal = 0,
+    ni,
+    headerLocation,
+    no,
+    na,
+    colName,
+    colType,
+    colKey,
+    annotations,
+    parsed
+
+  for (ni = 0; ni < chunks.length; ni++) {
+    annotations = {}
+    parsed = Papa.parse(csv.substring(chunks[ni].start, chunks[ni].stop)).data
+
+    headerLocation = 0
+    while (/^\s*#/.test(parsed[headerLocation][0])) {
+      headerLocation++
+    }
+
+    if (
+      parsed[headerLocation][1] === 'error' &&
+      parsed[headerLocation][2] === 'reference'
+    ) {
+      const ref = parsed[headerLocation + 1][2]
+      const msg = parsed[headerLocation + 1][1]
+
+      throw new Error(`[${ref}] ${msg}`)
+    }
+
+    for (no = 0; no < headerLocation; no++) {
+      annotations[parsed[no][0]] = parsed[no].reduce((p, c, i) => {
+        p[parsed[headerLocation][i]] = c
+        return p
+      }, {})
+    }
+
+    for (no = 1; no < parsed[headerLocation].length; no++) {
```

Comment:
why are we resetting `no` here instead of using a new variable? Kinda makes it confusing